File: masterClass/problemSolvingPatterns/anagrams.js
/* Given two strings, determine if one is an anagram of the other.

Examples:
validAnagram('','') // true
validAnagram('aaz','zza') // false
validAnagram('anagram','nagaram') // true
validAnagram('rat','car') // false
validAnagram('awesome','awesom') // false
*/

function validAnagram(str1, str2) {
  if (str1.length !== str2.length) return false;
  const list1 = {};
  for (let char of str1) {
    if (list1[char]) {
      list1[char]++;
    } else {
      list1[char] = 1;
    }
  }
  for (let char of str2) {
    if (list1[char]) {
      list1[char]--;
    } else {
      return false;
    }
  }
  return true;
}

validAnagram('anagram', 'nagaram');

File: masterClass/py/problem_patterns/freq_counter/anagrams.py
import collections

def valid_anagram(orig_str, possible_anag):
    """
    This function returns True if possible_anag is an anagram of orig_str.
    """
    return collections.Counter(orig_str) == collections.Counter(possible_anag)

print(valid_anagram("", "")) # true
print(valid_anagram("aaz", "zaa")) # true
print(valid_anagram("anagram", "nagaram")) # true
print(valid_anagram("rat", "car")) # false
print(valid_anagram("awesome", "awesom")) # false
print(valid_anagram("qwerty", "rwetqy")) # true
print(valid_anagram("texttwisttime", "timetwisttext")) #true